### ISSUE TITLE
Wrong parent snapshot with wildcard configuration

### ISSUE THREAD

**Issue by ilgiaco** [digint/btrbk#387]
Hi,
as discussed in other issues I set up wildcard backup for lxd containers backup. What I'm experiencing is that for some of the wildcard expanded targets, btrbk correctly creates the snapshots, but instead of copying it as initial backup, it tries to copy the snapshot using a snapshot from another container as parent, and it obviously fails.

As workaround I think I'll have to hardcode one by one every single container in btrbk.conf.

You can see attached btrbk.conf and btrbk.log
[btrbk.log](https://github.com/digint/btrbk/files/6462379/btrbk.log)
[btrbk.conf.txt](https://github.com/digint/btrbk/files/6462382/btrbk.conf.txt)

UPDATE
Hardcoding the targets one by one, I get the same result. The same containers fails trying to use another container's snapshot as parent, even commenting it out from btrbk configuration. I suspect there is something weird with some subvol identifier. Any suggestion?

**Comment by digint** [comment]
btrbk determines the best parent by a quite sophisticated algorithm:
```
  # preferences for parent (and required clone sources):
  #  1. closest older in snapdir (by btrbk timestamp), related
  #  2. closest older related (by cgen)
  #  3. closest newer related (by cgen)
  #  4. closest older in snapdir (by btrbk timestamp)
  #  5. closest newer in snapdir (by btrbk timestamp)
```

For some reasons, this does not seem optimal (or even plain wrong?) for your lxc setup.

please provide some trace logs for further analysis:
```
# btrbk run -n -l trace 2>&1 | grep -e get_best_parent -e related_nodes
```

**Comment by ilgiaco** [comment]
Hi Axel, thank you for your support!

here are my trace logs

[trace.log](https://github.com/digint/btrbk/files/6493811/trace.log)

**Comment by ilgiaco** [comment]
Ok, I think I have found something interesting.
After some analysis I found that all lxc root subvols have a parent_uuid (and I don't know why, probably something managed by lxd), all of these parent_uuids "do not exist", if I run ` btrfs subvolume show -u <parent_uuid> <root_btrfs_path>` I get 
```
ERROR: can't find uuid '<parent_uuid>' on '<root_btrfs_path>'
free(): double free detected in tcache 2
zsh: abort (core dumped)  btrfs subvolume show -u <parent_uuid> <root_btrfs_path>
```
BUT the most relevant thing is that some is that some lxc containers share the same parent, and I think this is the reason btrbk gets confused, and tries to mix containers snapshots with other containers' previous snapshot

**Comment by digint** [comment]
The missing `parent_uuid` should not be a problem, this is normal on btrfs file systems, for example:
```
# btrfs sub create A
Create subvolume './A'

# btrfs sub snap A B
Create a snapshot of 'A' in './B'

# lsbtr -u ./
MOUNT_SOURCE  ID   TOP  CGEN  GEN  UUID                                  PARENT_UUID                           RECEIVED_UUID  FLAGS  PATH
/dev/loop0    289  5    46    47   f79b0540-f753-7d45-9dda-000442a24dc4  -                                     -              -      /tmp/snaptest/A
/dev/loop0    290  5    47    47   7493ddcb-e206-354f-a60f-32dc3c3bfff2  f79b0540-f753-7d45-9dda-000442a24dc4  -              -      /tmp/snaptest/B

# btrfs sub delete A
Delete subvolume (no-commit): '/tmp/snaptest/A'

# lsbtr -u ./
MOUNT_SOURCE  ID   TOP  CGEN  GEN  UUID                                  PARENT_UUID                           RECEIVED_UUID  FLAGS     PATH
/dev/loop0    290  5    47    47   7493ddcb-e206-354f-a60f-32dc3c3bfff2  f79b0540-f753-7d45-9dda-000442a24dc4  -              -         /tmp/snaptest/B
```

I did not come around to analyze your logs yet (kind of forgot, and super busy...), do you still have this problem? What version of btrbk are you using?

**Comment by ilgiaco** [comment]
Hi, super busy period for me too.
I still have the problem, I think I've found the cause of it and a workaround.

### Cause
As I wrote in May (https://github.com/digint/btrbk/issues/387#issuecomment-849243407) the cause of my bug seems caused by the fact that lxd containers created with the same version of the same image, are created as a snapshot of the image, and so they share the same parent. So when btrbk tries to do an incremental backup for these new subvolumes and looks for nearest snapshot to use as parent snapshot, it gets as result the subvolume of the other container, and then it fails.

### Workaround
The workaround I found is to force complete backup the first time, setting `incremental no` in the btrbk.conf subvolume section. After the first backup I can remove the `incremental` setting and from that time onward it finds the correct parent and everything goes well.

If you need more infos and/or tests I'm at your disposition

btrbk version: 0.31.2

**Comment by digint** [comment]
Maybe this has something to do with the mount points, although btrbk should handle this correctly. Coincidentally I made the mount point mapping more robust yesterday, could you try latest master and report back if it helps?

> The workaround I found is to force complete backup the first time, setting incremental no in the btrbk.conf subvolume section.

There is also the `incremental_resolve` option to tell btrbk where to look for parents. Possibly setting `incremental_resolve directory` could help, but it theory it should be ok to look for parents on the mount point and not only in the configured snapshot_dir / target dirs.

From the [man page](/digint/btrbk/blob/master/doc/btrbk.conf.5.asciidoc#btrfs-specific-options):
> incremental_resolve mountpoint|directory
> Specifies where to search for the best common parent for incremental backups. If set to “mountpoint”, use parents in the filesystem tree below the mount point of the snapshot and target directory. If set to “directory”, use parents strictly below snapshot/target directories. Set this to “directory” if you get access problems (when not running btrbk as root). Defaults to “mountpoint”.

**Comment by digint** [comment]
Also, I pushed some code to the `alt-parent-selection` branch. This changes the algorithm used to pick the best parent.

From:
  1. closest older in snapdir (by btrbk timestamp), related
  2. closest older related (by cgen)
  3. closest newer related (by cgen)
  4. closest older in snapdir (by btrbk timestamp)
  5. closest newer in snapdir (by btrbk timestamp)

To:
1. closest older by btrbk timestamp in snapdir, with parent_uuid relationship
2. closest newer by btrbk timestamp in snapdir, with parent_uuid relationship
3. closest older by btrbk timestamp in snapdir, regardless of parent_uuid relationship
4. closest newer by btrbk timestamp in snapdir, regardless of parent_uuid relationship
5. closest older by cgen, with parent_uuid relationship
6. closest newer by cgen, with parent_uuid relationship

You can play around with it by altering the lines here:
https://github.com/digint/btrbk/blob/alt-parent-selection/btrbk#L3809

**Comment by digint** [comment]
incremental_prefs config option is now available in btrbk-0.32.0, closing issue
